File: src-tauri/src/launcher.rs
use crate::environment::{expand_path, get_prefix_path};
use crate::types::EnvironmentConfig;
use crate::types::WeGameStatus;
use anyhow::{Context, Result};
use chrono::Local;
use std::collections::HashMap;
use std::process::{Command, Stdio};
use tokio::sync::Mutex;

static WEGAME_PROCESS: Mutex<Option<u32>> = Mutex::const_new(None);

pub async fn launch_wegame(config: &EnvironmentConfig) -> Result<WeGameStatus> {
    let prefix_path = get_prefix_path(config);

    let wegame_exe = if config.wegame_install_path.is_empty() {
        prefix_path.join("drive_c/Program Files/Tencent/WeGame/WeGameLauncher.exe")
    } else {
        expand_path(&config.wegame_install_path).join("WeGameLauncher.exe")
    };

    if !wegame_exe.exists() {
        // Try alternative paths
        let alt_paths = [
            prefix_path.join("drive_c/Program Files (x86)/Tencent/WeGame/WeGameLauncher.exe"),
            prefix_path.join("drive_c/Program Files/Tencent/WeGame/WeGame.exe"),
        ];
        let found = alt_paths.into_iter().find(|p| p.exists());
        match found {
            Some(p) => return do_launch(&p, config).await,
            None => {
                return Err(anyhow::anyhow!(
                    "WeGame executable not found. Please install WeGame first.\nSearched in: {}",
                    wegame_exe.display()
                ));
            }
        }
    }

    do_launch(&wegame_exe, config).await
}

async fn do_launch(wegame_exe: &std::path::Path, config: &EnvironmentConfig) -> Result<WeGameStatus> {
    let proton_path = if config.proton_path.is_empty() {
        // Auto-detect
        let versions = super::proton::scan_proton_versions()?;
        super::proton::get_default_proton_path(&versions)
            .ok_or_else(|| anyhow::anyhow!("No Proton version found. Install GE-Proton first."))?
    } else {
        expand_path(&config.proton_path).to_string_lossy().to_string()
    };
    
    let proton_dir = std::path::Path::new(&proton_path)
        .parent()
        .ok_or_else(|| anyhow::anyhow!("Invalid Proton path"))?;

    // Build environment variables
    let mut env_map: HashMap<&str, String> = HashMap::new();
    env_map.insert("WINEPREFIX", get_prefix_path(config).to_string_lossy().to_string());
    env_map.insert("STEAM_COMPAT_CLIENT_INSTALL_PATH", get_prefix_path(config).to_string_lossy().to_string());
    env_map.insert("STEAM_COMPAT_DATA_PATH", get_prefix_path(config).parent().unwrap_or(&get_prefix_path(config)).to_string_lossy().to_string());
    env_map.insert("STEAM_RUNTIME_LIBRARY_PATH", "".to_string());
    env_map.insert("PROTON_LOG", "".to_string());
    env_map.insert("DISPLAY", ":0".to_string());
    env_map.insert("SDL_VIDEO_DONT_ALLOW_VULCAN", "1".to_string());

    // Merge extra env vars from config
    for (k, v) in &config.extra_env_vars {
        env_map.insert(k.as_str(), v.clone());
    }

    let wegame_str = wegame_exe.to_string_lossy().to_string();

    let mut cmd = Command::new(&proton_path);
    cmd.arg("run")
        .arg(&wegame_str);
        
    if !config.launch_args.is_empty() {
        cmd.arg(&config.launch_args);
    }

    cmd.envs(env_map.iter().map(|(k, v)| (*k, v.as_str())))
        .current_dir(wegame_exe.parent().unwrap_or(wegame_exe))
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

    let child = cmd.spawn().context("Failed to start WeGame process")?;
    let pid = child.id();

    // Store PID globally
    {
        let mut guard = WEGAME_PROCESS.lock().await;
        *guard = Some(pid);
    }

    let now = Local::now().format("%Y-%m-%d %H:%M:%S").to_string();
    
    Ok(WeGameStatus {
        running: true,
        pid: Some(pid),
        start_time: Some(now),
        uptime_seconds: Some(0),
    })
}

pub async fn stop_wegame() -> Result<WeGameStatus> {
    let pid = {
        let guard = WEGAME_PROCESS.lock().await;
        *guard
    };

    match pid {
        Some(pid) => {
            #[cfg(target_os = "linux")]
            {
                // Kill the process tree
                let _ = Command::new("kill")
                    .arg("-TERM")
                    .arg(pid.to_string())
                    .output();

                // Also kill any remaining WeGame-related processes
                let _ = Command::new("pkill")
                    .args(["-f", "WeGame"])
                    .output();
                
                // Wait a moment then force kill if still alive
                tokio::time::sleep(tokio::time::Duration::from_secs(2)).await;
                let _ = Command::new("kill")
                    .arg("-9")
                    .arg(pid.to_string())
                    .output();
            }

            {
                let mut guard = WEGAME_PROCESS.lock().await;
                *guard = None;
            }

            Ok(WeGameStatus {
                running: false,
                pid: None,
                start_time: None,
                uptime_seconds: None,
            })
        }
        None => Ok(WeGameStatus::default()),
    }
}

pub async fn check_wegame_status() -> WeGameStatus {
    let pid = {
        let guard = WEGAME_PROCESS.lock().await;
        *guard
    };

    match pid {
        Some(pid) => {
            #[cfg(target_os = "linux")]
            {
                let output = Command::new("kill")
                    .arg("-0")
                    .arg(pid.to_string())
                    .output();

                if output.is_ok_and(|o| o.status.success()) {
                    WeGameStatus {
                        running: true,
                        pid: Some(pid),
                        start_time: None,
                        uptime_seconds: None,
                    }
                } else {
                    let mut guard = WEGAME_PROCESS.lock().await;
                    *guard = None;
                    WeGameStatus::default()
                }
            }
            #[cfg(not(target_os = "linux"))]
            {
                WeGameStatus { running: true, pid: Some(pid), start_time: None, uptime_seconds: None }
            }
        }
        None => WeGameStatus::default(),
    }
}

File: electron/backend/launcher.ts
import { spawn } from "child_process";
import { EnvironmentConfig, WeGameStatus } from "./types";
import { expandPath, getPrefixPath } from "./environment";
import { scanProtonVersions, getDefaultProtonPath } from "./proton";
import path from "path";
import fs from "fs";
import { launcherLogger as log } from "./logger";

let wegameProcessPid: number | null = null;
let wegameStartTime: string | null = null;

export async function launchWegame(
  config: EnvironmentConfig
): Promise<WeGameStatus> {
  const prefixPath = getPrefixPath(config);

  let wegameExe: string;
  if (config.wegame_install_path) {
    wegameExe = path.join(
      expandPath(config.wegame_install_path),
      "WeGameLauncher.exe"
    );
  } else {
    wegameExe = path.join(
      prefixPath,
      "drive_c/Program Files/Tencent/WeGame/WeGameLauncher.exe"
    );
  }

  if (!fs.existsSync(wegameExe)) {
    // Try alternative paths
    const altPaths = [
      path.join(
        prefixPath,
        "drive_c/Program Files (x86)/Tencent/WeGame/WeGameLauncher.exe"
      ),
      path.join(
        prefixPath,
        "drive_c/Program Files/Tencent/WeGame/WeGame.exe"
      ),
    ];
    const found = altPaths.find((p) => fs.existsSync(p));
    if (found) {
      wegameExe = found;
    } else {
      throw new Error(
        `WeGame executable not found. Please install WeGame first.\nSearched in: ${wegameExe}`
      );
    }
  }

  log.info(`WeGame executable resolved: ${wegameExe}`);
  return doLaunch(wegameExe, config);
}

async function doLaunch(
  wegameExe: string,
  config: EnvironmentConfig
): Promise<WeGameStatus> {
  log.separator();
  log.info("=== WeGame Launch Start ===");
  log.info(`WeGame exe: ${wegameExe}`);
  log.info(`Config: ${JSON.stringify(config, null, 2)}`);

  let protonPath: string;
  if (config.proton_path) {
    protonPath = expandPath(config.proton_path);
    log.info(`Using user-specified Proton: ${protonPath}`);
  } else {
    const versions = scanProtonVersions();
    log.info(`Scanned Proton versions: ${versions.map(v => v.name).join(", ") || "(none)"}`);
    const defaultPath = getDefaultProtonPath(versions);
    if (!defaultPath) {
      log.error("No Proton version found!");
      throw new Error(
        "No Proton version found. Install GE-Proton first."
      );
    }
    protonPath = defaultPath;
    log.info(`Auto-selected Proton: ${protonPath}`);
  }

  const prefixPath = getPrefixPath(config);
  const prefixParent = path.dirname(prefixPath);

  // Build environment variables
  const env: Record<string, string> = {
    ...process.env as Record<string, string>,
    WINEPREFIX: prefixPath,
    STEAM_COMPAT_CLIENT_INSTALL_PATH: prefixPath,
    STEAM_COMPAT_DATA_PATH: prefixParent,
    STEAM_RUNTIME_LIBRARY_PATH: "",
    PROTON_LOG: "",
    DISPLAY: ":0",
    SDL_VIDEO_DONT_ALLOW_VULCAN: "1",
  };

  // Merge extra env vars from config
  for (const [k, v] of Object.entries(config.extra_env_vars)) {
    env[k] = v;
  }

  const args = ["run", wegameExe];
  if (config.launch_args) {
    args.push(config.launch_args);
  }

  log.info(`Spawn command: ${protonPath} ${args.join(" ")}`);
  log.info(`Working directory: ${path.dirname(wegameExe)}`);
  log.info(`Environment variables: WINEPREFIX=${env.WINEPREFIX}, STEAM_COMPAT_DATA_PATH=${env.STEAM_COMPAT_DATA_PATH}`);

  const child = spawn(protonPath, args, {
    cwd: path.dirname(wegameExe),
    env,
    stdio: ["ignore", "pipe", "pipe"],
    detached: true,
  });

  wegameProcessPid = child.pid ?? null;
  wegameStartTime = new Date().toISOString().replace("T", " ").slice(0, 19);

  log.info(`Process spawned with PID: ${wegameProcessPid}`);

  // Capture stdout/stderr to log file
  child.stdout?.on("data", (data: Buffer) => {
    for (const line of data.toString().split("\n")) {
      const trimmed = line.trim();
      if (trimmed) log.info(`[stdout] ${trimmed}`);
    }
  });

  child.stderr?.on("data", (data: Buffer) => {
    for (const line of data.toString().split("\n")) {
      const trimmed = line.trim();
      if (trimmed) log.warn(`[stderr] ${trimmed}`);
    }
  });

  child.on("close", (code) => {
    log.info(`WeGame process exited with code: ${code}`);
    wegameProcessPid = null;
    wegameStartTime = null;
  });

  child.on("error", (err) => {
    log.error(`WeGame process error: ${err.message}`);
  });

  // Don't wait for the child to exit
  child.unref();

  return {
    running: true,
    pid: wegameProcessPid ?? undefined,
    start_time: wegameStartTime ?? undefined,
    uptime_seconds: 0,
  };
}

export async function stopWegame(): Promise<WeGameStatus> {
  if (wegameProcessPid) {
    try {
      // Kill the process tree
      spawn("kill", ["-TERM", wegameProcessPid.toString()]);

      // Also kill any remaining WeGame-related processes
      spawn("pkill", ["-f", "WeGame"]);

      // Wait then force kill
      await new Promise((resolve) => setTimeout(resolve, 2000));
      spawn("kill", ["-9", wegameProcessPid.toString()]);
    } catch {
      // ignore
    }

    wegameProcessPid = null;
    wegameStartTime = null;
  }

  return {
    running: false,
    pid: undefined,
    start_time: undefined,
    uptime_seconds: undefined,
  };
}

export function checkWegameStatus(): WeGameStatus {
  if (!wegameProcessPid) {
    return { running: false };
  }

  try {
    // Check if process is still alive (signal 0 doesn't kill, just checks)
    process.kill(wegameProcessPid, 0);
    return {
      running: true,
      pid: wegameProcessPid,
      start_time: wegameStartTime ?? undefined,
      uptime_seconds: undefined,
    };
  } catch {
    // Process no longer exists
    wegameProcessPid = null;
    wegameStartTime = null;
    return { running: false };
  }
}
